internal/controller/numaflowcontrollerrollout_controller.go

```diff
@@ -192,42 +215,186 @@ func (r *NumaflowControllerRolloutReconciler) reconcile(
 		r.customMetrics.DecNumaflowControllerMetrics(controllerRollout.Name, controllerRollout.Namespace, controllerRollout.Spec.Controller.Version)
-		return nil
+		return ctrl.Result{}, nil
 	}
 
 	// add Finalizer so we can ensure that we take appropriate action when CRD is deleted
 	if !controllerutil.ContainsFinalizer(controllerRollout, finalizerName) {
 		controllerutil.AddFinalizer(controllerRollout, finalizerName)
 	}
 
-	// apply controller
+	_, pauseRequestExists := GetPauseModule().getControllerPauseRequest(namespace)
+	if !pauseRequestExists {
+		// this is just creating an entry in the map if it doesn't already exist
+		GetPauseModule().newControllerPauseRequest(namespace)
+	}
+
+	deployment, deploymentExists, err := r.getNumaflowControllerDeployment(ctx, controllerRollout)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if deploymentExists {
+		numaLogger.Debugf("found existing numaflow-controller Deployment")
+
+		if common.DataLossPrevention {
+
+			// if I need to update or am in the middle of an update of the Controller Deployment, then I need to make sure all the Pipelines are pausing
+			controllerDeploymentNeedsUpdating, controllerDeploymentIsUpdating, err := r.isControllerDeploymentUpdating(ctx, controllerRollout, deployment)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			numaLogger.Debugf("controllerDeploymentNeedsUpdating=%t, controllerDeploymentIsUpdating=%t", controllerDeploymentNeedsUpdating, controllerDeploymentIsUpdating)
+
+			// set the Status appropriately to "Pending" or "Deployed"
+			// if controllerDeploymentNeedsUpdating - this means there's a mismatch between the desired NumaflowControllerRollout version and actual
+			// if there's a generation mismatch - this means we haven't even observed the current generation
+			// we may match the first case and not the second when we've observed the generation change but we're pausing pipelines
+			// we may match the second case and not the first if we need to update something other than Numaflow Controller version
+			if controllerDeploymentNeedsUpdating || controllerRollout.Status.ObservedGeneration < controllerRollout.Generation {
+				controllerRollout.Status.MarkPending()
+			} else {
+				controllerRollout.Status.MarkDeployed(controllerRollout.Generation)
+			}
+
+			if controllerDeploymentNeedsUpdating || controllerDeploymentIsUpdating {
+				numaLogger.Info("Numaflow Controller either needs to or is in the process of updating")
+
+				// request pause if we haven't already
+				pauseRequestUpdated, err := r.requestPipelinesPause(ctx, controllerRollout, true)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+				if !pauseRequestUpdated { // don't do this yet if we just made a request - it's too soon for anything to have happened
+					// check if the pipelines are all paused and we're trying to update the spec
+					if controllerDeploymentNeedsUpdating {
```

Comment:
I think one nice thing about having these separate is that I can have a separate comment to explain why I have the condition `!pauseRequestUpdated`. Is there a reason you prefer to combine them?

Reply:
I prefer them to be combined to avoid to much nesting and indentation (for readability). You could always specify in the comment that the comment only applies to one of the 2 variables.

Reply:
okay, I'm fine to modify it